cpp/arcticdb/version/local_versioned_engine.cpp

```diff
@@ -1768,20 +1768,61 @@ timestamp LocalVersionedEngine::latest_timestamp(const std::string& symbol) {
 }
 
 std::unordered_map<KeyType, std::pair<size_t, size_t>> LocalVersionedEngine::scan_object_sizes() {
-    std::unordered_map<KeyType, std::pair<size_t, size_t>> sizes;
+    std::unordered_map<KeyType, std::pair<size_t, std::atomic<size_t>>> sizes;
     foreach_key_type([&store=store(), &sizes=sizes](KeyType key_type) {
-        std::vector<std::pair<VariantKey, stream::StreamSource::ReadContinuation>> keys_and_continuations;
-        auto& pair = sizes[key_type];
-        store->iterate_type(key_type, [&keys_and_continuations, &pair](const VariantKey &&k) {
-            keys_and_continuations.emplace_back(
-                std::forward<const VariantKey>(k),
-                [&pair](auto &&ks) {
-                    auto key_seg = std::move(ks);
-                    ++pair.first;
-                    pair.second += key_seg.segment().total_segment_size();
-                    return key_seg.variant_key();
-                });
+        std::vector<std::pair<VariantKey, stream::StreamSource::ReadContinuation>> keys;
+        store->iterate_type(key_type, [&keys, &sizes, &key_type](const VariantKey&& k) {
+            auto& pair = sizes[key_type];
+            ++pair.first;
+            auto& size_counter = pair.second;
+            keys.emplace_back(std::forward<const VariantKey>(k), [&size_counter] (auto&& ks) {
+                auto key_seg = std::move(ks);
+                size_counter += key_seg.segment().total_segment_size();
+                return key_seg.variant_key();
+            });
+        });
+
+        if(!keys.empty()) {
+            store->batch_read_compressed(std::move(keys), BatchReadArgs{}).get();
+        }
+
+    });
+
+    std::unordered_map<KeyType, std::pair<size_t, size_t>> result;
+    for (const auto& [k, v] : sizes) {
+        result[k] = v;
+    }
+    return result;
+}
+
+std::unordered_map<StreamId, std::unordered_map<KeyType, size_t>> LocalVersionedEngine::scan_object_sizes_by_stream() {
+    std::mutex mutex;
+    std::unordered_map<StreamId, std::unordered_map<KeyType, size_t>> sizes;
+    auto streams = symbol_list().get_symbols(store());
+
+    foreach_key_type([&store=store(), &sizes, &mutex](KeyType key_type) {
+        std::vector<std::pair<VariantKey, stream::StreamSource::ReadContinuation>> keys;
+
+        store->iterate_type(key_type, [&keys, &mutex, &sizes, key_type](const VariantKey&& k){
+            keys.emplace_back(std::forward<const VariantKey>(k), [key_type, &sizes, &mutex] (auto&& ks) {
+                auto key_seg = std::move(ks);
+                auto variant_key = key_seg.variant_key();
+                auto stream_id = variant_key_id(variant_key);
+                auto size = key_seg.segment().total_segment_size();
+
+                {
+                    std::lock_guard lock{mutex};
+                    sizes[stream_id][key_type] += size;
```

Comment:
Does it make sense to use atomic here as well? Or maybe switch to a spinlock instead of a mutex.

Reply:
This case is a bit different to the function above because we also need some synchronization guarantees over the map itself, so I don't think atomic on its own is safe. We could try spinlock, but really I'm not too worried - the locked operation should be lightning fast relative to the IO.

Reply:
> need some synchronization guarantees over the map itself

Is this because there can be a race on creating `sizes[stream_id]`?